share/smack/include/pthread.h

```diff
@@ -0,0 +1,71 @@
+//
```

Comment:
Allow me to try and summarize what you are saying:

1. There are many implementations of the pthread API, so we cannot depend on any particular implementation details, e.g., regarding the layout of the `pthread_mutex_t` structs.

2. The pthread API headers tend to include many other headers.

Please correct me if I am wrong.

For the first issue, I don’t see what the problem is. Why couldn’t we reuse existing pthread API headers, and then link in our own implementation of the pthread API? Why should we care how any particular system defines, e.g., `pthread_mutex_t`? We could treat it as an opaque pointer if we wanted to, or we could provide our own layout — like you are doing — no? Is there any reason why linking in our own implementation of the pthread API against the standard pthread headers would not succeed?

The only problem I can imagine is if pthread API clients were to break the API abstraction, e.g., by accessing platform-specific struct layouts. Does this happen often? My advice would be to suppose it does not really happen, and say that your implementation would be unsound if it did.

For the second issue, is it really a problem? As far as I am aware, including additional *code* — i.e., procedure implementations — can be costly at analysis time. What would go wrong if we were to include a whole bunch of additional *headers* though?

I think that sticking to original pthread headers is more practical and user friendly. If we were to write programs / projects from scratch, there is no problem to include your custom pthread API headers. However, I am guessing we want to analyze programs that are already written, and compile against some system headers. In that case compiling against your custom headers might fail, as you have already noticed via altering the include order.

I think the right thing to do is to try and avoid providing custom pthread API headers, avoid modifying the compilation process, and simply link our own pthread API implementation instead of the system’s.